sudo/lib/exec/parent.rs

```diff
@@ -212,15 +216,41 @@ impl ParentClosure {
             }
         }
     }
+
+    /// Handle changes to the monitor status.
+    fn handle_sigchld(&mut self) {
+        const OPTS: WaitOptions = WaitOptions::new().all().untraced().no_hang();
+
+        let (pid, status) = loop {
+            match waitpid(WaitPid::any(), OPTS) {
+                Err(WaitError::Io(err)) if was_interrupted(&err) => {}
+                Err(_) => return,
+                Ok(ok) => break ok,
+            }
+        };
+
+        if Some(pid) == self.monitor_pid {
+            if status.did_exit() || status.was_signaled() {
+                self.monitor_pid = None;
+            } else if status.was_stopped() {
+                // FIXME: we should stop too.
+            }
+        } else {
+            // FIXME: what other child could we have?
+        }
+    }
 }
 
 impl EventClosure for ParentClosure {
     type Break = ParentMessage;
 
-    fn on_signal(&mut self, info: SignalInfo, dispatcher: &mut EventDispatcher<Self>) {
+    fn on_signal(&mut self, info: SignalInfo, _dispatcher: &mut EventDispatcher<Self>) {
+        if self.monitor_pid.is_none() {
```

Comment:
maybe this could be `let Some(monitor_pid) = self.monitor_pid else { return }` and then you can pass that `monitor_pid` binding into `handle_sigchld`. following the previous idea of calling `waitpid(monitor_pid)` in `handle_sigchld`

Reply:
yes! I was thinking the same thing :)